test/integration/egress.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"istio.io/manager/test/util"
+)
+
+type egress struct {
+	*infra
+}
+
+func (t *egress) String() string {
+	return "egress proxy"
+}
+
+func (t *egress) setup() error {
+	return nil
+}
+
+func (t *egress) run() error {
+	if !t.Egress {
+		return nil
+	}
+	extServices := []string{
+		"httpbin",
+		//"httpsbin",TODO
+	}
+
+	funcs := make(map[string]func() status)
+	for _, src := range []string{"a", "b"} {
+		for _, dst := range extServices {
+			name := fmt.Sprintf("External request from %s to %s", src, dst)
+			funcs[name] = (func(src, dst string) func() status {
+				url := fmt.Sprintf("http://%s/headers", dst)
+				trace := fmt.Sprint(time.Now().UnixNano())
+				return func() status {
+					resp, err := util.Shell(fmt.Sprintf(
+						"kubectl exec %s -n %s -c app -- client -url %s -insecure -key Trace-Id -val %q",
+						t.apps[src][0], t.Namespace, url, trace))
+					if err != nil {
+						glog.Error(err)
```

Comment:
An error message is printed twice for this failure: once here (via glog.Error) and another by the caller of parallel(). It might be better to propagate the request status via normal error handling, e.g.

```golang
type errRequestRepeat struct {}
func (e *errRequestRepeat) Error() string { "" } // doesn't matter since we never print it
```
Then the request runner in parallel would look something like this:

```golang
if err := f(); err != nil {
    if _, ok := err.(errRequestRequest); ok {
        continue // try again
    }
    return  fmt.Errorf("Failed %s at attempt %d", name, n)
}
return nil // success
```